[Lotto - Step2] Submit miooo0o's mission

miooo0o opened the pull request:
Hello! Hyeon-Gu Choi,https:

I’m happy I managed to finish it somehow! I’m also grateful that I had the chance to grow with your feedback. Thank you so much again.

### What I focused on

- splitting`controller` as much as possible, by function(_tried_). 
  - `PurchasesController`
  - `Prompter:` handle ask/read/inputview
  - `LottoWinningCreator` 
  - `TicketsEvaluator` 
  - `TicketIssuer`
- created `data classes` to adopt new `controller`/s.
- removing `service` altogether. was helpful!
- reassemble the data with `data classes` while avoiding `mutable` state and `var`.



### What I am not happy -- most
- there are too many values in classes that should have been `private`... I made messy lines.

### What I couldn’t manage
- git branch
- I planned to refactor and expand the `Validator` logic but couldn’t complete that part.
- I tended to commit in larger chunks instead of splitting them properly.
- I didn’t fully stick to TDD /  README

---

### A small but big question about my process

While refactoring, I felt overwhelmed by my own code 😢. 
At some point, I couldn’t go back and just had to finish it, especially when wrapping classes.

On the other hand, I felt like there was nothing left to test for the new data classes. Because they were just wrapping other classes I’d already tested, so they seemed self-contained. 
But sure, was not: I didn’t realize I had actually removed a lot of validation during refactoring.

When I go with the flow/ overwhelmed like this, it’s easy to miss important parts. While coding alone, I didn’t have time to think about the README or feature list either.

As an engineer, do you have any advice for me on how to handle this better?

Thank you very much for your time!
I did my best. and it was super fun! 

재밌게 했습니다! 시간 내주셔서 정말 감사합니다!

Hyeon9mak commented on src/main/kotlin/lotto/controller/PurchasesController.kt:
```diff
@@ -0,0 +1,18 @@
+import lotto.dto.PurchaseResult
+import lotto.dto.UserPurchase
+
+object PurchasesController {
+    fun getPurchaseResult(): PurchaseResult {
+        val userPurchase = getUserPurchases()
+        return PurchaseResult(userPurchase)
+    }
+
+    private fun getUserPurchases(): UserPurchase {
+        val amount = Prompter.askAmount()
+        val manualTicketsCount = Prompter.askManualTicketsNumbers(amount)
```
I understood these controllers are separated!
They're get all DTO from user before in earnest domain logic, right?

miooo0o replied:
Yes, that's right!
I made it so that all user input is collected first.

Hyeon9mak commented on src/main/kotlin/lotto/controller/PurchasesController.kt:
```diff
@@ -0,0 +1,18 @@
+        val manualTicketsCount = Prompter.askManualTicketsNumbers(amount)
+        return UserPurchase(amount, manualTicketsCount)
+    }
+}
```
need check the EOF 🥹

miooo0o replied:
Oh no 😿,
I'll fix it. I just learned that Gradle has forced formatting...

Hyeon9mak commented on src/main/kotlin/lotto/controller/TicketsEvaluator.kt:
```diff
@@ -0,0 +1,18 @@
+        val totalPrize = ranks.sumOf { it.winningMoney }
+        return if (totalPrize == 0) 0.0 else (totalPrize.toDouble() / totalAmount) * 100.0
+    }
+}
```
i think it's not used 🤔

Hyeon9mak commented on src/main/kotlin/lotto/controller/TicketsEvaluator.kt:
```diff
@@ -0,0 +1,18 @@
+import lotto.domain.WinningCombination
+import lotto.dto.IssuedTickets
+
+object TicketsEvaluator {
+    fun runWith(issuedTickets: IssuedTickets, winningCombination: WinningCombination): List<Rank> {
+        return issuedTickets.get().map {
+            it.getRank(winningCombination)
+        }
+    }
```
I think the best to practice is how View, Controller, and Domain interact with each other,
but I still think this kind of attempt is valuable.

As you try out different approaches like this, I believe you'll naturally start thinking about better structures.

For now, I understand what you were aiming for, so be great to refine this a bit more!

Hyeon9mak commented on src/main/kotlin/lotto/domain/NumberGenerator.kt:
```diff
@@ -0,0 +1,11 @@
+package lotto.domain
+
+object NumberGenerator {
+    fun get(): Set<Int> {
+        return (LottoNumber.MIN_RANGE_NUMBER..LottoNumber.MAX_RANGE_NUMBER)
+            .shuffled()
+            .take(LottoTicket.SUFFICIENT_SIZE)
+            .sorted()
+            .toCollection(linkedSetOf())
+    }
+}
```
I guess the reason of made this object class, is make more easier to generate test code. 🤔 
isn't it?

miooo0o replied:
Oh, There was a similar naming method hidden in other file, I thought deleting 'it' so I would fix this problem. Sorry! I will write a new test for it! @Hyeon9mak

Hyeon9mak commented on src/main/kotlin/lotto/domain/WinningCombination.kt:
```diff
@@ -8,7 +8,7 @@ import lotto.exceptions.LottoException
 data class WinningCombination(val winningTicket: LottoTicket, val bonusNumber: LottoNumber) {
     init {
         require(
-            !winningTicket.get().contains(bonusNumber)
+            !winningTicket.contains(bonusNumber)
         ) { LottoException.InvalidBonusNumberException(bonusNumber.hashCode()) }
```
How about remake negative method?
it can remove `!` keyword.

```kotlin
require(winningTicket.doesNotContain(bounusNumber))
```

miooo0o replied:
That's right. This looks much better.
I'll have to figure out how to trim the conditions/or make readable.
```kotlin
require(winningTicket.doesNotContains(bonusNumber)) { ExceptionMessage.DUPLICATE }
```

Thanks.

Hyeon9mak commented on src/main/kotlin/lotto/dto/PurchaseResult.kt:
```diff
@@ -0,0 +1,11 @@
+
+import lotto.controller.TicketIssuer
+
+data class PurchaseResult(
+    val userPurchase: UserPurchase,
+) {
+    val manualTickets = TicketIssuer.issueManual(userPurchase.manualTicketsCount)
+    val randomTickets = TicketIssuer.issueRandom(userPurchase.randomTicketsCount)
+    val totalTickets = manualTickets + randomTickets
```
`manualTickets` & `randomTicket` are not used! 😮

miooo0o replied:
oh, true!

what do you think about,

```kotlin
    val totalTickets =
        TicketIssuer.issueManual(userPurchase.manualTicketsCount) +
            TicketIssuer.issueRandom(userPurchase.randomTicketsCount)
```

this?

miooo0o commented:
Hello. Hyeon9mak.

My comments were late this time! Thank you for your understanding.

I realized while doing Blackjack that I've been neglecting communication and not following basic practices well. 
Thank you for understanding my immature communication. You probably had a lot to check too, and I could have had a conversation with you before that. It's unfortunate that I chose to complete it and show you instead of using a better approach. And also, as you mentioned, I did make many changes before, but it wasn't focused on business and was more like personal practice. 

That aspect really hit me recently, so I wanted to express my gratitude to you.

---

With my last commits, somehow, there were minor-issue-things, but that's exactly why it feels that much critical.

The changes this time are as follows:

- Removed "unnecessary files" that I learned about during the review. There were really more than I thought.
- Added basic validation processes
I tried some things that I thought might be fun, but I'm just revert/delete and submitting the easiest and most understandable `object`. It's simply a collection of requires! I like it.
- Fixed errors in the formula for calculating total rate
- Checked all EOF (with ktlinFormat! )


thank you in advance, have a good day :)

Hyeon9mak commented on src/main/kotlin/lotto/dto/RankedTickets.kt:
```diff
@@ -0,0 +1,15 @@
+import lotto.controller.TicketsEvaluator
+import lotto.domain.WinningCombination
+
+data class RankedTickets(val purchaseResult: PurchaseResult, val winningCombination: WinningCombination) {
+    val ranked = TicketsEvaluator.runWith(purchaseResult.totalTickets, winningCombination)
+    val totalRate = getTotalRate()
+
+    private fun getTotalRate(): Double {
+        val totalPrize = ranked.sumOf { it.winningMoney }
+        val totalSpent = purchaseResult.userPurchase.amount
+        return totalPrize.toDouble() / totalSpent
+    }
+}
```
nooooooooooo. u need check about this getter method!
It makes compilation error 👿

miooo0o replied:
Ohhh how could I miss this? I should probably wait until I'm more alert before sending review requests. I tend to make silly mistakes when I do it in the evening.

Thank you

Hyeon9mak commented on src/main/kotlin/lotto/domain/NumberGenerator.kt:
```diff
@@ -1,11 +1,11 @@
 package lotto.domain
 
-object RandomNumbers {
+object NumberGenerator {
     fun get(): Set<Int> {
         return (LottoNumber.MIN_RANGE_NUMBER..LottoNumber.MAX_RANGE_NUMBER)
             .shuffled()
             .take(LottoTicket.SUFFICIENT_SIZE)
             .sorted()
             .toCollection(linkedSetOf())
     }
```
https://github.com/herotechcourse/kotlin-lotto/pull/28#discussion_r2195200863
Where is the test code ?! 👀 
I think u can generate test code for this wouldn’t be too hard. 😄

miooo0o replied:
I tired!

Hyeon9mak commented on src/main/kotlin/lotto/domain/LottoNumber.kt:
```diff
@@ -1,6 +1,6 @@
 package lotto.domain
 
-data class LottoNumber private constructor(private val value: Int) {
+data class LottoNumber(private val value: Int) {
     companion object {
```
and now, i think u can apply the `value class` 😄 
https://github.com/herotechcourse/kotlin-lotto/pull/8#discussion_r2193347343

miooo0o replied:
yeah. good point. I tried....

```kotlin
@JvmInline
value class LottoNumber(private val value: Int) {
    companion object {
        internal const val MIN_RANGE_NUMBER = 1
        internal const val MAX_RANGE_NUMBER = 45
        private val LOTTO_NUMBERS = (MIN_RANGE_NUMBER..MAX_RANGE_NUMBER).associateWith(::LottoNumber)

        fun from(request: Int): LottoNumber {
            return LOTTO_NUMBERS[request] ?: throw IllegalArgumentException()
        }
    }

    override fun toString(): String = value.toString()
}
```

like this. But I could not pass the test, 

```kotlin
        assertThat(LottoNumber.from(MIN_RANGE_NUMBER)).isSameAs(number)
```

in my understanding, `value class` not confirm that returning same instance. so I let them be as `data class`.
What do you think?

Hyeon9mak commented on src/main/kotlin/lotto/domain/LottoTicket.kt:
```diff
@@ -1,16 +1,25 @@
+@JvmInline
+value class LottoTicket(private val lottoNumbers: HashSet<LottoNumber>) {
     init {
-        require(lottoNumbers.size == SUFFICIENT_SIZE)
+        require(lottoNumbers.size == SUFFICIENT_SIZE) to ExceptionMessage.NOT_SUFFICIENT_SIZE
     }
```
Have u checked what kind of exception message is shown when an error occurs? 🤔 
U can check this by test code.

```ktolin
@Test
    fun `throw exception message, when lotto ticket has not enough or dubplicate numbers`() {
        val exception = assertThrows<IllegalArgumentException> {
            LottoTicket(
                hashSetOf(
                    LottoNumber.from(1),
                    LottoNumber.from(2),
                    LottoNumber.from(3),
                    LottoNumber.from(4),
                    LottoNumber.from(5),
                )
            )
        }
        assertThat(exception.message).isEqualTo("lotto ticket has not sufficient size")
    }
```

miooo0o replied:
Oh no way, I thought I caught all the exceptions but totally missed this one.
I'm missing way more stuff than I realized. Thank you!

Hyeon9mak reviewed (changes requested):
Hi Mina, nice job working through the feedback.
Unfortunately, there’s a compilation error in the current version of the code, which makes it impossible to run the program or the tests. 😢 

Just like a chef checks for hygiene and food readiness, or an architect ensures safety and structural integrity.
A software engineer should also make sure the basics are working before deploying — it’s a core part of our craft!

I’ve left just a few simple comments.
Keep going, you’re almost there!
And as always, feel free to reach out if you have any questions 😄

miooo0o commented:
Today I'm leaving a short message as I'm feeling low on energy and very embarrassed haha.

#### changed
- Made it possible to test random numbers
- Changed LottoNumber to a value class
- Removed as many potential sources of critical errors as possible
- Sending the request after completing basic tests I hope this works well and there are no issues this time.

#### What was challenging: Changed LottoNumber to a value class
When I tried to change something I had been using without much thought, there were more things to change than I expected, and I learned this time that runtime errors could occur due to code I left behind without removing (trying to read HashCode), which was good to know.

Thank you always.

Hyeon9mak commented on src/main/kotlin/lotto/domain/LottoNumber.kt:
```diff
@@ -1,6 +1,11 @@
 package lotto.domain
 
-data class LottoNumber private constructor(private val value: Int) {
+@JvmInline
+value class LottoNumber(private val value: Int) {
+    fun getValue(): Int = value
+
+    override fun toString(): String = value.toString()
+
```
> in my understanding, value class not confirm that returning same instance. so I let them be as data class.
What do you think?

<img width="1140" height="628" alt="image" src="https://github.com/user-attachments/assets/90a9bcfe-7e8c-48b5-ac14-e4e0925c8331" />

🤔 
need read the official docs! 
https://kotlinlang.org/docs/inline-classes.html

miooo0o replied:
I'll read it again. I still don't understand much of what the document says. I was wondering about the equal/same test with value class. Perhaps I need to study more. Thank you :)

Hyeon9mak commented on src/main/kotlin/lotto/domain/LottoNumber.kt:
```diff
@@ -1,6 +1,11 @@
 package lotto.domain
 
-data class LottoNumber private constructor(private val value: Int) {
+@JvmInline
+value class LottoNumber(private val value: Int) {
+    fun getValue(): Int = value
+
+    override fun toString(): String = value.toString()
```
```suggestion
@JvmInline
value class LottoNumber(val value: Int) {

    init {
        // require ...
    }
    
    companion object {
        internal const val MIN_RANGE_NUMBER = 1
        internal const val MAX_RANGE_NUMBER = 45
    }
}
```

u can change like this.
(Jason gave a nice hint!) cause value class change original data type at compile time!

miooo0o replied:
Hi! Very nice... advice! I applied it to Lotto step2 and then to all other missions. 
I didn't need to use getter, but I did it out of habit! Thanks to Jason too!

Hyeon9mak reviewed (approved):
Hi Mina, nice work in Lotto mission.
I'm approving this PR, but there’s still one point I really hope you’ll revisit.

There is a case where the program still doesn't work properly.
I think it's important to take time to check each of these and refine the implementation accordingly.

<img width="326" height="289" alt="image" src="https://github.com/user-attachments/assets/fbd6bc69-4d9b-4da3-9632-8e3945173d3c" />

Setting aside OOP, Clean Code, and TDD — as professional engineers, delivering a program that works correctly is the most basic responsibility.
Just like how we'd be upset if a brand-new phone we paid a lot for didn’t work as expected.

I believe good habits make great engineers. I hope you don’t overlook these small but important things.
I trust you'll take even more care in the next missions!
